models/search.js

```diff
@@ -0,0 +1,146 @@
+import database from 'infra/database';
+import validator from 'models/validator.js';
+
+async function doSearch(values) {
+  values = validateValues(values);
+
+  const scopes = {
+    contents: searchContent,
+  };
+
+  const scopeSearch = scopes[values.search_scope];
+  return await scopeSearch(values);
+
+  function validateValues(values) {
+    const cleanValues = validator(values, {
+      page: 'optional',
+      per_page: 'optional',
+      search_scope: 'required',
+      search_term: 'required',
+    });
+
+    return cleanValues;
+  }
+}
+
+async function searchContent(inputValues) {
+  const results = {};
+  let query = {};
+
+  inputValues.page = inputValues.page || 1;
+  const offset = (inputValues.page - 1) * inputValues.per_page;
+
+  query.values = [
+    inputValues.per_page || 30,
+    offset,
+    null,
+    'published',
+    getSearchWordsFromSearchTerm(inputValues.search_term),
+  ];
+
+  query.text = `WITH content_window AS (
+      SELECT
+        COUNT(*) OVER()::INTEGER as total_rows,
+        id
+      FROM contents
+      WHERE contents.parent_id IS NOT DISTINCT FROM $3 AND contents.status = $4 AND (to_tsvector('portuguese', unaccent(title)) @@ to_tsquery('portuguese', unaccent($5)) OR to_tsvector('portuguese', unaccent(body)) @@ to_tsquery('portuguese', unaccent($5)))
+      ORDER BY ts_rank(to_tsvector(contents.body),to_tsquery($5)) DESC NULLS LAST, published_at DESC
```

Comment:
Isso aqui também pode ser bastante complexo se for ter que lidar com todos os conteúdos que derem match. Lembrando que o limite e offset só são aplicados depois de ordenar.

Reply:
Esse é um excelente ponto também! Fiz alguns testes sem utilizar a ordenação mas o resultado N fica muito legal... dessa forma ele apresenta os conteúdos de acordo com a sua relevância em relação ao que foi buscado... no teste automatizado tem um exemplo bem bom disso onde podemos ver bem o funcionamento dessa ordenação... será que teria alguma alternativa?

Reply:
Lendo alguns foruns na internet, peguei uma ideia de fazer uma ordenação condicional... para pesquisas de apenas uma palavra mantive a ordenação por ordem de publicação, já que o rank dificilmente traria resultados de ordenação diferentes, assim a gente poupa um pouco as buscas,

Uma outra ideia que eu tive é fazer uma busca um pouco mais acertiva... ao invés de adicionar termos com OR, adicionaria com AND... assim reduziria o número de conteúdos filtrados... a contrapartida é que diminuiria o escopo de busca apenas para conteúdos que contenham todas as palavras digitadas no search term... ficaria um pouco menos amplo do que está atualmente... oq acha sobre isso?